### ISSUE TITLE
Body class containes _ instead of -

### ISSUE THREAD

**Issue by ksz2013** [magento/magento2#19225]
### Summary (*)
Body classes generation has changed.

### Examples (*)
For the latest 2.3 version in this commit https://github.com/magento/magento2/commit/00fc06d4466e9d2ab61714f4741224b3b016d57a#diff-991e58213a3fece7cb59fd74cb824f8f

The unescape for _ for the body classes was removed. Is this necessary?
With this change if i have a controller Catalog/Product/View inside my Namespace/Modulename 
the body class which was generated in this format: **_namespace-modulename-catalog-product-view_** becomes
**_namespace_modulename-catalog_product-view_**

Until this commit there was impossible to have body classes with _, all _'s were converted to -.
This commit breaks that logic.

### Proposed solution
Can that commit be reverted back or fixed.

PS: 
Or change it to:

$className = preg_replace('#[^a-z0-9-]+#', '-', strtolower($className));

Regards,
Szabi

**Comment by magento-engcom-team** [comment]
Hi @ksz2013. Thank you for your report. 
To help us process this issue please make sure that you provided the following information:
- [ ] Summary of the issue
- [ ] Information on your environment
- [ ] Steps to reproduce
- [ ] Expected and actual results

Please make sure that the issue is reproducible on the vanilla Magento instance following **Steps to reproduce**. To deploy vanilla Magento instance on our environment, please, add a comment to the issue: 

`@magento-engcom-team give me $VERSION instance` 

where `$VERSION` is [version tags](https://github.com/magento/magento2/tags) (starting from 2.2.0+) or develop branches (for example: 2.3-develop). 
For more details, please, review the [Magento Contributor Assistant](https://devdocs.magento.com/guides/v2.2/contributor-guide/contributing.html#contributor-assist) documentation.

@ksz2013 do you confirm that you was able to reproduce the issue on vanilla Magento instance following steps to reproduce?
- [ ] yes
- [ ] no

**Comment by rusdragos** [comment]
I'm interested in this as well, it is breaking functionality in many of our extensions. Is this change really necessary?

**Comment by DanielRuf** [comment]
> The unescape for _ for the body classes was removed.

I can not see this in the linked diff. Can you point us to the right change?
Also this commit just adds support for BEM classes.

**Comment by ksz2013** [comment]
Hi @DanielRuf ,

https://github.com/magento/magento2/commit/00fc06d4466e9d2ab61714f4741224b3b016d57a#diff-9715340011b3760b01c96d997e7e7894R466

Basically here it was changed 

$className = preg_replace('#[^a-z0-9]+#', '-', strtolower($className));

to

$className = preg_replace('#[^a-z0-9-_]+#', '-', strtolower($className));

But this leaves body classes with _ in them such as namespace_modulename-catalog_product-view
instead namespace-modulename-catalog-product-view

From what i see,  if they change it to: 
$className = preg_replace('#[^a-z0-9-]+#', '-', strtolower($className));

it would fix the problem, and also offer support to BEM classes. (the _ from the regexp should be removed)

Regards

**Comment by magento-engcom-team** [comment]
Hi @engcom-backlog-nazar. Thank you for working on this issue. 
In order to make sure that issue has enough information and ready for development, please read and check the following instruction: :point_down:
- [x] 1. Verify that issue has all the required information. (**Preconditions**, **Steps to reproduce**, **Expected result**, **Actual result**).<details><summary>Details</summary>If the issue has a valid description, the label `Issue: Format is valid` will be added to the issue automatically. Please, edit issue description if needed, until label `Issue: Format is valid` appears.</details>

- [x] 2. Verify that issue has a meaningful description and provides enough information to reproduce the issue. If the report is valid, add `Issue: Clear Description` label to the issue by yourself.

- [x] 3. Add `Component: XXXXX` label(s) to the ticket, indicating the components it may be related to.
- [ ] 4. Verify that the issue is reproducible on `2.3-develop` branch<details><summary>Details</summary>- Add the comment `@magento-engcom-team give me 2.3-develop instance` to deploy test instance on Magento infrastructure. <br/> - If the issue is reproducible on `2.3-develop` branch, please, add the label `Reproduced on 2.3.x`.<br/>- If the issue is not reproducible, add your comment that issue is not reproducible and close the issue and **_stop verification process here_**!
</details>

- [ ] 5. Verify that the issue is reproducible on `2.2-develop` branch. <details><summary>Details</summary>- Add the comment `@magento-engcom-team give me 2.2-develop instance` to deploy test instance on Magento infrastructure. <br/>- If the issue is reproducible on `2.2-develop` branch, please add the label `Reproduced on 2.2.x`</details>

- [ ] 6. Add label `Issue: Confirmed` once verification is complete. 

- [ ] 7. Make sure that automatic system confirms that report has been added to the backlog.

**Comment by ghost** [comment]
hi @ksz2013 thank you for you report, this cannot be reverted, this commit fixes another issue -> https://github.com/magento/magento2/issues/10645

**Comment by mattwellss** [comment]
Isn't this a backwards-compatibility break? For example, the CustomerBalance page `baseurl/storecredit/info/` included the body class `magento-customerbalance-info-index`, whereas in `2.2.9` (and `2.3.2`, I presume) the class is `magento_customerbalance-info-index`.

it seems like a valid issue, and it's caused problems for us both in core modules and custom ones. The class where the change was made is marked as `@api`, so stability is supposedly promised, unless I'm misunderstanding the system.
